### PR TITLE
Swoc cache tool

### PR DESCRIPTION
**Pull request by persiaAziz-zz**

### CODE REVIEW

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 1]
```diff
@@ -521,6 +534,30 @@ Stripe::dir_freelist_length(int s)
...
+int
+Stripe::check_loop(int s)
+{
+  // look for loop in the segment
+  // rewrite the freelist if loop is present
+  int free           = 0;
+  CacheDirEntry *seg = this->dir_segment(s);
+  CacheDirEntry *e   = dir_from_offset(this->freelist[s], seg);
+  std::bitset<65536> f_bitset;
+  f_bitset.reset();
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:714]: I think this is unnecessary - the bitset should initialize to all 0.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 2]
```diff
@@ -521,6 +534,30 @@ Stripe::dir_freelist_length(int s)
   return free;
 }
 
+int
+Stripe::check_loop(int s)
+{
+  // look for loop in the segment
+  // rewrite the freelist if loop is present
+  int free           = 0;
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:710]: What is `free` for?

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 3]
```diff
@@ -521,6 +534,30 @@ Stripe::dir_freelist_length(int s)
   return free;
 }
 
+int
+Stripe::check_loop(int s)
+{
+  // look for loop in the segment
+  // rewrite the freelist if loop is present
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:709]: Isn't this just the detection part?

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 4]
```diff
@@ -521,6 +534,30 @@ Stripe::dir_freelist_length(int s)
...
+  int free           = 0;
+  CacheDirEntry *seg = this->dir_segment(s);
+  CacheDirEntry *e   = dir_from_offset(this->freelist[s], seg);
+  std::bitset<65536> f_bitset;
+  f_bitset.reset();
+  while (e) {
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:715]: Why not
```
int i = this->freelist[s];
while (nullptr != (e = dir_from_offset(i, seg)))
```

File: cmd/traffic_cache_tool/CacheTool.cc (hunk 3 above)
Comment (persiaAziz-zz) [reply cmd/traffic_cache_tool/CacheTool.cc:709]: yes it is. This command is just a note for myself.  I have the rewrite part in dir_init_segment.

File: cmd/traffic_cache_tool/CacheDefs.h [hunk 5]
```diff
@@ -42,6 +42,25 @@ using namespace ApacheTrafficServer;
 
 namespace ts
 {
+#define dir_clear(_e) \
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheDefs.h:46]: Don't use `#define` unless absolutely required. Use inline functions. That avoid argument evaluation issues and bizarre things like the `do` / `while(0)` construct.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 6]
```diff
@@ -426,6 +432,103 @@ dir_to_offset(const CacheDirEntry *d, const CacheDirEntry *seg)
 #endif
 }
 
+
+CacheDirEntry *
+Stripe::dir_delete_entry(CacheDirEntry *e, CacheDirEntry *p, int s)
+{
+  CacheDirEntry *seg         = dir_segment(s);
+  int no           = dir_next(e);
+  this->_meta[0][0].dirty = 1;
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: Is it always copy A? Also, don't use numbers here. Use `A` or `B` and `HEAD` or `FOOT`.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 7]
```diff
@@ -426,6 +432,103 @@ dir_to_offset(const CacheDirEntry *d, const CacheDirEntry *seg)
...
+  if (p) {
+    unsigned int fo = this->freelist[s];
+    unsigned int eo = dir_to_offset(e, seg);
+    dir_clear(e);
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:510]: For Later: These need to be moved to methods, not free functions.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 8]
```diff
@@ -426,6 +432,103 @@ dir_to_offset(const CacheDirEntry *d, const CacheDirEntry *seg)
...
 
+
+CacheDirEntry *
+Stripe::dir_delete_entry(CacheDirEntry *e, CacheDirEntry *p, int s)
+{
+  CacheDirEntry *seg         = dir_segment(s);
+  int no           = dir_next(e);
+  this->_meta[0][0].dirty = 1;
+  if (p) {
+    unsigned int fo = this->freelist[s];
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc]: Some comments might be nice...it's fine for us, now, but we must make this code such that other people have a chance of understanding it. Let's not inflict on others the pain we've had trying to figure out what's going on.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 9]
```diff
@@ -426,6 +432,103 @@ dir_to_offset(const CacheDirEntry *d, const CacheDirEntry *seg)
...
+      return nullptr;
+    }
+  }
+  return dir_from_offset(no, seg);
+}
+
+void
+Stripe::walk_all_buckets()
+{
+  for (int s = 0; s < this->_segments; s++) {
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:534]: `++s`.

File: cmd/traffic_cache_tool/CacheTool.cc [hunk 10]
```diff
@@ -456,6 +559,16 @@ Stripe::dir_init_segment(int s)
...
 }
 
+Errata
+Stripe::loadDir()
+{
+  Errata zret;
+  char *raw_dir = (char *)ats_memalign(ats_pagesize(), this->vol_dirlen());
+  dir           = (CacheDirEntry *)(raw_dir + this->vol_headerlen());
+  // read directory
+  pread(this->_span->_fd, raw_dir, this->vol_dirlen(), this->_start);
```
Comment (SolidWallOfCode) [inline cmd/traffic_cache_tool/CacheTool.cc:635]: Check the return of `pread`? Especially for the CacheTool, it must check *everything*.

**Review by SolidWallOfCode — CHANGES_REQUESTED** [review]

File: cmd/traffic_cache_tool/CacheDefs.h (hunk 5 above)
Comment (persiaAziz-zz) [reply cmd/traffic_cache_tool/CacheDefs.h:46]: Some of the functions here are implemented in the the exact same way as ATS for debugging purposes. I will reimplement these after I am done tracking the outstanding bugs.

**Review by SolidWallOfCode — APPROVED** [review]

**Pull request merged**